### ISSUE TITLE
Devcontainer for vscode.dev?

### ISSUE THREAD

**Issue by lattice0** [microsoft/vscode-remote-release#9059]
Hi, I connected to a tunnel and I cannot use devcontainer or even install the Dev Containers extension on vscode.dev. I'm trying to code completely on the cloud but devcontainer is very much needed for me because I always destroy my VMs doing lota of stuff and I don't want to reinstall stuff.

Why there's no support for dev containers on tunnels? 

Why it works on codespaces online if it's simply an online vscode?

**Comment by lzunsec** [comment]
currently cannot do remote development because this is missing, please add 🙏

**Comment by jon77p** [comment]
Similar to my comment in #7705 and original issue in #7715 nearly a year ago, but this support would be crucial to have to fully enable development on vscode.dev using personal infrastructure separate from the Codespaces workflow.

Devcontainers are a great feature to apply reusability/repeatability concepts towards development environments, so it would be great to have this feature parity across all ways of accessing and using the vscode environment.

**Comment by 77see** [comment]
need this on my ipad and also our company can move to cloud development finally

**Comment by jjcasmar** [comment]
Is there any update on being able to connect to devcontainer from vscode.dev?

**Comment by lattice0** [comment]
@jjcasmar 
It looks like it's in the backlog. I think we need to raise awareness so microsoft moves to in progress soon. I'm dying for this feature, already on my limit on github codespaces. Maybe send this thread to some friends

**Comment by jjcasmar** [comment]
If at least it was possible to open a tunnel directly to a devcontainer, instead of to the host...

**Comment by lattice0** [comment]
@jjcasmar it is, you can run a devcontainer and then run vscode tunnel installer as a post install script or on the dockerfile, but it would be a little annoying and not the way it should be done

**Comment by jjcasmar** [comment]
Yeah, but its not possible to do it automatically because it requires authentication. 
https://github.com/microsoft/vscode-remote-release/issues/9132
https://github.com/microsoft/vscode-remote-release/issues/9131

**Comment by Ilykuleshov** [comment]
I'd like to echo the need for this, as this would also allow a zero-installation full-featured coding solution on a browser-based OS, such as Chromium OS, e.g. for coding on low-grade/old laptops.

**Comment by lattice0** [comment]
@Ilykuleshov same here, on our company we want people to code on whatever device they have, even tablets (Tab S9 Ultra here). Codespaces is cool but too expensive and also for things like Android development where you need to have USB connection or maybe a local wireless connection for installing the APK on the phone over ADB, it's better to use a nearby machine. 

Without devcontainers though it's not easy to switch between development VMs easily (yes, we use VMs for isolation, and then containers for building with the correct dependencies)

**Comment by neet-ev** [comment]
this is the only feature missing for us to use vscode on our company

**Comment by matt-swanson** [comment]
+1 hopefully this allows attaching to docker containers as well.

**Comment by lattice0** [comment]
I tried his workaround of doing a live share from a devcontainer but vscode.dev refuses to load it https://github.com/microsoft/vscode-remote-release/issues/9493

I want this featureee so baaaad

**Comment by YoRyan** [comment]
I didn't want to wait for Microsoft to work on this, so I wrote my own [program](https://github.com/YoRyan/myspace). It builds dev containers using the dev container CLI and then injects the code CLI to create a tunnel. It isn't quite seamless, since you need to authenticate every time you create a new tunnel, and you only get five slots on vscode.dev. But it includes support for devcontainer.json-defined settings and extensions, so it almost feels as good as Codespaces.

Until such time as there is official support, it does the job for me.

**Comment by lattice0** [comment]
My current solution is to stream VSCode from linux with [Sunshine](https://github.com/LizardByte/Sunshine) and use an Android Moonlight client. This is sad.

Is VSCode simply allowed the local VSCode to open in a browser, then I could TCP forward from the tablet, which would be even better as I don't have to pass through Microsoft servers which adds delay.

**Comment by YoRyan** [comment]
There's now a `code serve-web` command that starts a self-hosted, standalone VS Code server that looks and performs great. I have this command implemented in myspace as the `local` subcommand, and it's *almost* perfect for the DIY Codespaces use case. But unfortunately, the dev container CLI does not yet implement port forwarding for dev containers (https://github.com/devcontainers/cli/issues/186). That means we only get one port (appPort) which we have to use for the VS Code server itself. Without access to the other ports, you can't reach any services you spin up in your dev container. :cry:

**Comment by lattice0** [comment]
@YoRyan first of all, thanks, just tested this solution and is much faster via local ssh forwarding over my local network. Unfortunately the devcontainer plugin still does not work. 

For the devcontainer CLI, you're starting one with VSCode installed and then accessing it from outside? It's a good workaround but I wanted so bad for this to work with the extension already

**Comment by YoRyan** [comment]
> For the devcontainer CLI, you're starting one with VSCode installed and then accessing it from outside?

Precisely. The devcontainer CLI is made by Microsoft, so in theory the containers it creates should be just like those created by the official extension, or Codespaces.

Obviously it would be preferable to have first-party support, but the extension is closed-source, so the ball is firmly in Microsoft's court.

**Comment by worm991** [comment]
i need this feature for coding with default environments with no setup when someone joins our company, we use GPUs so cloud is needed but web version is essential

**Comment by iguanus** [comment]
I am also looking for a fully functional web interface. Tailscale + VSCode dev server works fine with native VS Code installations.

However, one cannot do this with a Quest 3 device, as one can't install Tailscale nor VSCode there. So the online interface allows for truly remote development, bypassing the client device issue. Since it works through SSH, I don't see the impediment to use it through the TCP tunnel, other than the lack of implementation on Microsoft's side.

**Comment by grutin111** [comment]
need this to code on my home server with much more processing power, without burning my laptop, and also allocate one VM for an employee, but without the hassle of setting everything. Devcontainers on web is a must

**Comment by OriLifschitz** [comment]
This is really needed, please please please add!

**Comment by worm991** [comment]
this is seriously preventing me from use VSCode! Extremely needed! Please Microsoft its been 1 year and there are solutions from Google and Zed coming...

**Comment by he500** [comment]
needed in our company asap

**Comment by starball5** [comment]
Related on Stack Overflow: [How to use a devcontainer in vscode.dev?](https://stackoverflow.com/q/77338169/11107541)

**Comment by rutilism** [comment]
is microsoft avoiding this for us to pay for codespaces? this is annoying, we need for local android development, it's not even supported on codespaces

there's no way weŕe going to install all C++ dependencies, android JDK, flutter, plugins, on all computers. We have a devcontainer for this already.

**Comment by chrmarti** [comment]
We hear you! This is just non-trivial to do because the Dev Containers extension so far assumes that it can use Node.js API and the local filesystem and these parts of the code will need to change before we can make the extension available in the browser.

**Comment by he500** [comment]
for malware research this would also be useful, no local vscode installation with possible vulnerabilities, just the browser connecting to a VM with all the tools on the devcontainer. Also for developing using less known plugins

**Comment by adrianer-asr** [comment]
any news on this? I tried and it still wont work

**Comment by 8192K** [comment]
> We hear you! This is just non-trivial to do because the Dev Containers extension so far assumes that it can use Node.js API and the local filesystem and these parts of the code will need to change before we can make the extension available in the browser.

@chrmarti Why not create a new extension that deals with devcontainers through tunnels thus not having to rely on the limitations of or having to wait for changes in the existing devcontainer extension?

**Comment by jjcasmar** [comment]
Is there any ETA for this feature?

**Comment by lattice0** [comment]
2 years already and no updates

**Comment by serphen** [comment]
I have implemented my version of that feature here: https://github.com/serphen/remote-devcontainers

<img width="943" height="614" alt="Image" src="https://github.com/user-attachments/assets/0d002ebc-8436-4bc7-ba1c-60833f8b87fe" />

----
When Dev Containers needs something, on the web, it doesn't  know whom to speak to. So I developed a new extension that says *"oh ok, you can ask me if you need anything"*.
It's on top of Dev Containers.

There is no trick to it, in VS Code there is an official API where an extension can register what it can handles, so the extension claims that it can handle `dev-container+<hex>` (the same mechanism Microsoft uses on the desktop).

In turn, when you do an action, this new extension asks the machine that is actually hosting Docker: *"I need you to boot this docker, stream me the log files, etc, etc."* 

This is a complete separate pipeline of the one that Microsoft did.

You can quickly understand once you see the repo: there is a whole new Docker orchestrator and support for port forwarding.  
Obviously AI (Claude and Kimi K3) was massively used to write the host/orchestrator side because the scope is quite large (dozens of features) and it's really not that straightforward at all (spent almost a week on this).

So, if you are brave enough and like unofficial stuff, you can now use dev containers remotely.

I did it for fun, so it may be incomplete (or surprisingly good). In the UI, host is blue, containers are purple.

**Comment by tdegrunt** [comment]
Would this enable going from vscode.dev, using a tunnel to my home Mac and there using devcontainers?

**Comment by 8192K** [comment]
Why does run.sh download VS Code? Also, how can I enable this extension using vscode.dev on Safari on iPadOS? Thanks.
